### PR TITLE
fix(iot-dev): reuse IotHubConnection objects on reconnection #395

### PR DESCRIPTION
**Pull request by jasmineymlo**

### CODE REVIEW

**Comment by jasmineymlo** [comment]
Preview for the fix and e2e tests addition if any will be added in the next iteration.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 1]
```diff
@@ -164,6 +166,21 @@ public void addDeviceOperationSession(DeviceClientConfig deviceClientConfig) thr
         }
     }
 
+    public void reset(ScheduledExecutorService scheduledExecutorService) throws TransportException
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java]: >reset [](start = 16, length = 5)

Could this logic be moved to the open call instead? That way, we only have to close, then open the xxxiothubconection objects

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java]: if we can, we would also need to move some of the constructor logic into the open call, too

---
In reply to: [258633143](https://github.com/Azure/azure-iot-sdk-java/pull/445#discussion_r258633143) [](ancestors = 258633143)

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 2]
```diff
@@ -112,7 +109,7 @@ public AmqpsIotHubConnection(DeviceClientConfig config, ScheduledExecutorService
             throw new IllegalArgumentException("hubName cannot be null or empty.");
         }
 
-        this.scheduledExecutorService = scheduledExecutorService;
+        //this.scheduledExecutorService = scheduledExecutorService;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java]: can this be removed?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 3]
```diff
@@ -194,8 +169,20 @@ public void reset(ScheduledExecutorService scheduledExecutorService) throws Tran
...
+        reconnectionScheduled = false;
+        connectionId = UUID.randomUUID().toString();
+
+        this.scheduledExecutorService = scheduledExecutorService;
+
+        this.closeLatch = new CountDownLatch(1);
+        this.openLatch = new CountDownLatch(1);
+        this.savedException = null;
+
+        this.state = IotHubConnectionStatus.DISCONNECTED;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java]: Shouldn't this method check the state of the connection object before setting it to DISCONNECTED? What if it was already connected here?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java (hunk 2 above)
Comment (jasmineymlo) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java]: fixed

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java (hunk 3 above)
Comment (jasmineymlo) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java]: fixed

**Pull request merged**